src/core/server/saved_objects/export/saved_objects_exporter.ts

```diff
@@ -150,27 +161,113 @@ export class SavedObjectsExporter {
+
+    const findWithPit = async ({ id, searchAfter }: { id: string; searchAfter?: unknown[] }) => {
+      return await this.#savedObjectsClient.find({
+        // Sort fields are required to use searchAfter, so we set some defaults here
+        sortField: 'updated_at',
+        sortOrder: 'desc',
+        pit: {
+          keepAlive: '1m', // bump keep_alive by 1m on every new request
+          ...findOptions.pit,
+          id,
+        },
+        ...(searchAfter ? { searchAfter } : {}),
+        ...findOptions,
+      });
+    };
+
+    // Open PIT and request our first page of hits
+    let { id: pitId } = await this.#savedObjectsClient.openPointInTimeForType(findOptions.type);
+    let results = await findWithPit({ id: pitId });
+
+    let lastResultsCount = results.saved_objects.length;
+    let lastHitSortValue = getLastHitSortValue(results);
+    pitId = results.pit_id!;
+
+    this.#log.debug(`Collected [${lastResultsCount}] saved objects for export.`);
+
+    // Close PIT if this was our last page
+    if (lastResultsCount < findOptions.perPage!) {
+      await this.#savedObjectsClient.closePointInTime(pitId);
+    }
+
+    yield results;
+
+    // We've reached the end when there are fewer hits than our perPage size
+    while (lastHitSortValue && lastResultsCount === findOptions.perPage) {
+      results = await findWithPit({
+        id: pitId,
+        searchAfter: lastHitSortValue,
+      });
+
+      lastResultsCount = results.saved_objects.length;
+      lastHitSortValue = getLastHitSortValue(results);
+      pitId = results.pit_id!;
+
+      this.#log.debug(`Collected [${lastResultsCount}] more saved objects for export.`);
+
+      if (lastResultsCount < findOptions.perPage) {
+        await this.#savedObjectsClient.closePointInTime(pitId);
+      }
+
+      yield results;
+    }
+  }
+
   private async fetchByTypes({
     types,
     namespace,
     hasReference,
     search,
   }: SavedObjectsExportByTypeOptions) {
-    const findResponse = await this.#savedObjectsClient.find({
+    const options: SavedObjectsFindOptions = {
       type: types,
       hasReference,
       hasReferenceOperator: hasReference ? 'OR' : undefined,
       search,
-      perPage: this.#exportSizeLimit,
       namespaces: namespace ? [namespace] : undefined,
-    });
-    if (findResponse.total > this.#exportSizeLimit) {
-      throw SavedObjectsExportError.exportSizeExceeded(this.#exportSizeLimit);
+      // We aren't using `exportSizeLimit` here because a user may opt to set it
+      // higher than the 10k ES default for `index.max_result_window`, in which
+      // case we will use PIT to "scroll" through pages of hits, 10k at a time.
+      perPage: 10000,
+    };
+    const finder = this.findWithPointInTime(options);
+
+    const hits: SavedObjectsFindResult[] = [];
+    for await (const result of finder) {
+      hits.push(...result.saved_objects);
+      if (hits.length > this.#exportSizeLimit) {
```

Comment:
Some nits which we can also circle back to after FF:

1. We don't seem to have any documentation for this option, so it would be good to add that.
2. We might want to mention the configuration option or link to the docs from the error message https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/export/errors.ts#L30
3. Now that we support > 10k exports we might want to bump this up to a larger value, as long as it doesn't consume so much memory that Kibana becomes unstable. It will depend a lot on the size of the specific saved objects, but maybe we can test if e.g. 20k objects would really be a problem.

Reply:
Agreed on the above - would prefer to circle back to it as I also noticed we are incorrectly validating this config as bytes, when it should simply be a number. It'd be nice to fix all of those things at once if possible.